tiledb/ml/readers/pytorch.py

```diff
@@ -87,17 +94,45 @@ def __iter__(self) -> Iterator[DataType]:
             iter_start = worker_id * per_worker
             iter_end = min(iter_start + per_worker, rows)
 
+        offsets = range(iter_start, iter_end, self.batch_size)
+
+        # Shuffle offsets in case we need batch shuffling
+        if self.batch_shuffle:
+            random.seed()
+            offsets = random.sample(  # type: ignore
+                range(iter_start, iter_end, self.batch_size),
+                math.ceil((iter_end - iter_start) / self.batch_size),
+            )
+
         # Loop over batches
         with ThreadPoolExecutor(max_workers=2) as executor:
-            for offset in range(iter_start, iter_end, self.batch_size):
+            for offset in offsets:
                 x_batch, y_batch = run_io_tasks_in_parallel(
                     executor, (self.x, self.y), self.batch_size, offset
                 )
 
-                # Yield the next training batch
-                yield tuple(x_batch[attr] for attr in self.x_attribute_names) + tuple(
-                    y_batch[attr] for attr in self.y_attribute_names
-                )
+                if self.within_batch_shuffle:
+                    random.seed()
+                    # We get batch length based on the first attribute, because last batch might be smaller than the
+                    # batch size
+                    rand_permutation = np.arange(
+                        x_batch[self.x_attribute_names[0]].shape[0]
+                    )
+                    random.shuffle(rand_permutation)
```

Comment:
Numpy has its own random lib, `np.random`; for consistency either use `np.arange` with `np.random` or `list(range(..))` with `random`.